concrete-core/src/specification/engines/glwe_ciphertext_ggsw_ciphertext_discarding_external_product.rs

```diff
@@ -0,0 +1,81 @@
+use crate::specification::engines::AbstractEngine;
+use crate::specification::entities::{GgswCiphertextEntity, GlweCiphertextEntity};
+
+use super::engine_error;
+
+engine_error! {
+    GlweCiphertextGgswCiphertextDiscardingExternalProductError for
+    GlweCiphertextGgswCiphertextDiscardingExternalProductEngine @
+    PolynomialSizeMismatch => "All the GGSW and GLWE ciphertexts polynomial sizes must be the same.",
+    GlweDimensionMismatch => "All the GGSW and GLWE ciphertexts GLWE dimension must be the same."
+}
+
+impl<EngineError: std::error::Error>
+    GlweCiphertextGgswCiphertextDiscardingExternalProductError<EngineError>
+{
+    /// Validates the inputs
+    pub fn perform_generic_checks<GlweCiphertext, GgswCiphertext>(
+        glwe_input: &GlweCiphertext,
+        ggsw_input: &GgswCiphertext,
+        output: &GlweCiphertext,
+    ) -> Result<(), Self>
+    where
+        GlweCiphertext: GlweCiphertextEntity,
+        GgswCiphertext: GgswCiphertextEntity,
+    {
+        if glwe_input.polynomial_size().0 != ggsw_input.polynomial_size().0
+            || glwe_input.polynomial_size().0 != output.polynomial_size().0
+        {
+            return Err(Self::PolynomialSizeMismatch);
+        }
+        if glwe_input.glwe_dimension().0 != ggsw_input.glwe_dimension().0
+            || glwe_input.glwe_dimension().0 != output.glwe_dimension().0
+        {
+            return Err(Self::PolynomialSizeMismatch);
+        }
+        Ok(())
+    }
+}
+
+/// A trait for engines computing the external product between a GLWE ciphertext
+/// and a GSW ciphertext.
+///
+/// # Semantics
+///
+/// This [discarding](super#operation-semantics) operation fills the `output` GLWE ciphertext with
+/// the result of the external product between a `glwe_input` GLWE ciphertext and
+/// a `ggsw_input` GSW ciphertext.
+///
+/// # Formal Definition
+/// GSW(M) = (Glev(-SM), Glev(M))
+/// GLWE(\Delta M) = (A, B) (k = 1)
+/// GSW(M_1) \odot GLWE(\Delta M_2) = <Dec(B), Glev(M_1)> + <Dec(A), Glev(-SM_1)>
+/// = GLWE(\Delta M_1 M_2)
```

Comment:
could you please remove the formal def (leave the `# Formal Defiition` line) because it will be taken care of later with the crypto doc.